docs/RFCS/sql_batching.md

```diff
@@ -0,0 +1,499 @@
+- [+] Fits a nice mental model/easy to reason about
+- [+] Does not need to be opt-in, although we probably still want it to be, at least
+  at first
+- [+] More generally applicable
+- [+] Could support batched reads
+- [-] May not be possible to use with ORMs (or may at least require custom handling)
+- [-] Inflexible with regard to allowing users to decide which statements get executed
+  in parallel
+
+### Conversational API Support
+
+In contrast to sending a list of statements separated by semicolons, most SQL clients
+talk to databases using a single statement at-a-time conversational API. This means that
+the clients send a statement and wait for the response before proceeding. They expect at the
+point of a response for a statement that the statement has already been applied. An
+alternate approach to batching statements for semicolon-separated lists is to build
+batches between statement declarations when clients use this more standard conversational
+API. To do so, we would need to buffer statements instead of executing them immediately
+when we decide (or are told) to batch statements. We would then mock out a successful response
+value to immediately send back to the client. After this, the `sql.Executor` would wait
+for the next statement in the conversation to decide whether to continue building up a
+collection of statements or to execute all pending statements. The decision to either execute
+the accumulated statements or add to the batch would come down to if the new statement was
+[independent](#independent-statements) from the previous set of statements. Additionally,
+we would always need to execute the pending batch on read queries and transaction commit
+statements.
+
+To perform statement batching using this interface, users would need to specify which
+statements could be batched. They would do so using some new syntax, for instance appending
+`RETURNING NOTHING` to the end of INSERT and UPDATE statements. In turn, adding this
+syntax would indicate that the SQL executor could return a fake response (`RowsAffected = 1`)
+for this statement immediately and that the client would not expect the result to have any
+real meaning. A list of pros and cons to this approach are listed below:
+
+- [+] Fits the communication model ORMs usually work with
+- [+] Fits the communication model other SQL clients usually work with
+- [+] Users get more control over which statements are executed in a batch
+- [-] Requires us to mock out results. This will not be possible for more complicated
+  mutations that contain clauses like `RETURNING <values>`. It is an open question how
+  we should mock a response for statements like UPDATE, that are supposed to return the
+  number of rows affected
+- [-] The mocking out of results also means that the statement batching feature would be
+  unusable for any situation where the real statement result is needed
+- [-] Complicates error handling tremendously. How do we associate errors with the
```

Comment:
I don't think this is complicated. It's just a feature of concurrency batching: if an error occurs in any of the concurrent branches, you only get to know about it at the final synchronization, and you may lose the knowledge of where the error came from. The "early response" is not a mark of success -- it merely means "result not known yet".

We can make this slightly less inconvenient (although I personally think it's not inconvenient to start with) by 1) prefixing the error message with the statement that caused it (or at least the statement tag). 2) introducing a new statement "CHECK" (name does not matter) which is always considered as dependent, and which has no other purpose than complete/execute the current batch and check for any pending errors.